File: python_preprocessing/average_psd_max_channels.py
import h5py
import hdf5storage
import numpy as np

theta = np.zeros((2, 24, 4, 20))
alpha = np.zeros((2, 24, 4, 20))
lowbeta = np.zeros((2, 24, 4, 20))
highbeta = np.zeros((2, 24, 4, 20))
lowgamma = np.zeros((2, 24, 4, 20))
highgamma = np.zeros((2, 24, 4, 20))

for i in range(20):
    f=h5py.File('..\CLAS_VP{:0>2d}_onedata_STBFH_MNT.mat'.format(i+1),'r')
    allgoodevents=f['allgoodevents']
    allevents=f['allevents']
    new_1=[]
    new_2=[]
    for j in range(80):
        if(allgoodevents[j,0]==1):
            if(allevents[j,0]==13):
                new_1.append(j)
            else:
                new_2.append(j)
    
    f=h5py.File('..\stft_VP{:0>2d}_psd_con1.mat'.format(i+1),'r')
    psd=f['psd']
    new_psd_1=np.zeros((len(new_1),128,24,501))
    new_psd_2=np.zeros((len(new_2),128,24,501))
    for j in range(len(new_1)):
        new_psd_1[j,:,:,:]=psd[new_1[j],:,:,:]
    for j in range(len(new_2)):
        new_psd_2[j,:,:,:]=psd[new_2[j],:,:,:]
    mean_psd_1=np.mean(new_psd_1,0)
    mean_psd_2=np.mean(new_psd_2,0)
    for k in range(24):
        psdInTheta=np.reshape(mean_psd_1[j,:,k,4:8],(128,4))
        eachMax=np.amax(psdInTheta,0)
        eachCh=np.argmax(psdInTheta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        theta[0,k,0,i]=bandMax
        theta[1,k,0,i]=maxCh
        psdInAlpha=np.reshape(mean_psd_1[j,:,k,8:13],(128,5))
        eachMax=np.amax(psdInAlpha,0)
        eachCh=np.argmax(psdInAlpha,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        alpha[0,k,0,i]=bandMax
        alpha[1,k,0,i]=maxCh
        psdInLowbeta=np.reshape(mean_psd_1[j,:,k,14:21],(128,7))
        eachMax=np.amax(psdInLowbeta,0)
        eachCh=np.argmax(psdInLowbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowbeta[0,k,0,i]=bandMax
        lowbeta[1,k,0,i]=maxCh
        psdInHighbeta=np.reshape(mean_psd_1[j,:,k,20:31],(128,11))
        eachMax=np.amax(psdInHighbeta,0)
        eachCh=np.argmax(psdInHighbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highbeta[0,k,0,i]=bandMax
        highbeta[1,k,0,i]=maxCh
        psdInLowgamma=np.reshape(mean_psd_1[j,:,k,30:61],(128,31))
        eachMax=np.amax(psdInLowgamma,0)
        eachCh=np.argmax(psdInLowgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowgamma[0,k,0,i]=bandMax
        lowgamma[1,k,0,i]=maxCh
        psdInHighgamma=np.reshape(mean_psd_1[j,:,k,60:101],(128,41))
        eachMax=np.amax(psdInHighgamma,0)
        eachCh=np.argmax(psdInHighgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highgamma[0,k,0,i]=bandMax
        highgamma[1,k,0,i]=maxCh
        
        psdInTheta=np.reshape(mean_psd_2[j,:,k,4:8],(128,4))
        eachMax=np.amax(psdInTheta,0)
        eachCh=np.argmax(psdInTheta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        theta[0,k,1,i]=bandMax
        theta[1,k,1,i]=maxCh
        psdInAlpha=np.reshape(mean_psd_2[j,:,k,8:13],(128,5))
        eachMax=np.amax(psdInAlpha,0)
        eachCh=np.argmax(psdInAlpha,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        alpha[0,k,1,i]=bandMax
        alpha[1,k,1,i]=maxCh
        psdInLowbeta=np.reshape(mean_psd_2[j,:,k,14:21],(128,7))
        eachMax=np.amax(psdInLowbeta,0)
        eachCh=np.argmax(psdInLowbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowbeta[0,k,1,i]=bandMax
        lowbeta[1,k,1,i]=maxCh
        psdInHighbeta=np.reshape(mean_psd_2[j,:,k,20:31],(128,11))
        eachMax=np.amax(psdInHighbeta,0)
        eachCh=np.argmax(psdInHighbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highbeta[0,k,1,i]=bandMax
        highbeta[1,k,1,i]=maxCh
        psdInLowgamma=np.reshape(mean_psd_2[j,:,k,30:61],(128,31))
        eachMax=np.amax(psdInLowgamma,0)
        eachCh=np.argmax(psdInLowgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowgamma[0,k,1,i]=bandMax
        lowgamma[1,k,1,i]=maxCh
        psdInHighgamma=np.reshape(mean_psd_2[j,:,k,60:101],(128,41))
        eachMax=np.amax(psdInHighgamma,0)
        eachCh=np.argmax(psdInHighgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highgamma[0,k,1,i]=bandMax
        highgamma[1,k,1,i]=maxCh

for i in range(20):
    f=h5py.File('..\CLAS_VP{:0>2d}_onedata_STBFH_MNT.mat'.format(i+1),'r')
    allgoodevents=f['allgoodevents']
    allevents=f['allevents']
    new_1=[]
    new_2=[]
    for j in range(80):
        if(allgoodevents[j,0]==1):
            if(allevents[j,0]==13):
                new_1.append(j)
            else:
                new_2.append(j)
    
    f=h5py.File('..\stft_VP{:0>2d}_psd_con2.mat'.format(i+1),'r')
    psd=f['psd']
    new_psd_1=np.zeros((len(new_1),128,24,501))
    new_psd_2=np.zeros((len(new_2),128,24,501))
    for j in range(len(new_1)):
        new_psd_1[j,:,:,:]=psd[new_1[j],:,:,:]
    for j in range(len(new_2)):
        new_psd_2[j,:,:,:]=psd[new_2[j],:,:,:]
    mean_psd_1=np.mean(new_psd_1,0)
    mean_psd_2=np.mean(new_psd_2,0)
    
    for k in range(24):
        psdInTheta=np.reshape(mean_psd_1[j,:,k,4:8],(128,4))
        eachMax=np.amax(psdInTheta,0)
        eachCh=np.argmax(psdInTheta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        theta[0,k,2,i]=bandMax
        theta[1,k,2,i]=maxCh
        psdInAlpha=np.reshape(mean_psd_1[j,:,k,8:13],(128,5))
        eachMax=np.amax(psdInAlpha,0)
        eachCh=np.argmax(psdInAlpha,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        alpha[0,k,2,i]=bandMax
        alpha[1,k,2,i]=maxCh
        psdInLowbeta=np.reshape(mean_psd_1[j,:,k,14:21],(128,7))
        eachMax=np.amax(psdInLowbeta,0)
        eachCh=np.argmax(psdInLowbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowbeta[0,k,2,i]=bandMax
        lowbeta[1,k,2,i]=maxCh
        psdInHighbeta=np.reshape(mean_psd_1[j,:,k,20:31],(128,11))
        eachMax=np.amax(psdInHighbeta,0)
        eachCh=np.argmax(psdInHighbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highbeta[0,k,2,i]=bandMax
        highbeta[1,k,2,i]=maxCh
        psdInLowgamma=np.reshape(mean_psd_1[j,:,k,30:61],(128,31))
        eachMax=np.amax(psdInLowgamma,0)
        eachCh=np.argmax(psdInLowgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowgamma[0,k,2,i]=bandMax
        lowgamma[1,k,2,i]=maxCh
        psdInHighgamma=np.reshape(mean_psd_1[j,:,k,60:101],(128,41))
        eachMax=np.amax(psdInHighgamma,0)
        eachCh=np.argmax(psdInHighgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highgamma[0,k,2,i]=bandMax
        highgamma[1,k,2,i]=maxCh
        
        psdInTheta=np.reshape(mean_psd_2[j,:,k,4:8],(128,4))
        eachMax=np.amax(psdInTheta,0)
        eachCh=np.argmax(psdInTheta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        theta[0,k,3,i]=bandMax
        theta[1,k,3,i]=maxCh
        psdInAlpha=np.reshape(mean_psd_2[j,:,k,8:13],(128,5))
        eachMax=np.amax(psdInAlpha,0)
        eachCh=np.argmax(psdInAlpha,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        alpha[0,k,3,i]=bandMax
        alpha[1,k,3,i]=maxCh
        psdInLowbeta=np.reshape(mean_psd_2[j,:,k,14:21],(128,7))
        eachMax=np.amax(psdInLowbeta,0)
        eachCh=np.argmax(psdInLowbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowbeta[0,k,3,i]=bandMax
        lowbeta[1,k,3,i]=maxCh
        psdInHighbeta=np.reshape(mean_psd_2[j,:,k,20:31],(128,11))
        eachMax=np.amax(psdInHighbeta,0)
        eachCh=np.argmax(psdInHighbeta,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highbeta[0,k,3,i]=bandMax
        highbeta[1,k,3,i]=maxCh
        psdInLowgamma=np.reshape(mean_psd_2[j,:,k,30:61],(128,31))
        eachMax=np.amax(psdInLowgamma,0)
        eachCh=np.argmax(psdInLowgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        lowgamma[0,k,3,i]=bandMax
        lowgamma[1,k,3,i]=maxCh
        psdInHighgamma=np.reshape(mean_psd_2[j,:,k,60:101],(128,41))
        eachMax=np.amax(psdInHighgamma,0)
        eachCh=np.argmax(psdInHighgamma,0)
        bandMax=np.amax(eachMax)
        index=np.argmax(eachMax)
        maxCh=eachCh[index]
        highgamma[0,k,3,i]=bandMax
        highgamma[1,k,3,i]=maxCh
            
hdf5storage.savemat('..\Channels_mean_max_psd.mat', {'theta':theta, 'alpha':alpha, 'lowbeta':lowbeta, 'highbeta':highbeta, 'lowgamma':lowgamma,'highgamma':highgamma})

File: average_psd_max_channels.m
%计算每个记忆任务下所有trials的功率谱密度（psd）平均值，然后选出5个频率段上平均能量最大的通道
theta = zeros(2, 24, 4, 20);
alpha = zeros(2, 24, 4, 20);
lowbeta = zeros(2, 24, 4, 20);
highbeta = zeros(2, 24, 4, 20);
lowgamma = zeros(2, 24, 4, 20);
highgamma = zeros(2, 24, 4, 20);
for i=1:20
    load(['stft_VP', sprintf('%02d', i), '_psd_con1.mat']);
    load(['CLAS_VP', sprintf('%02d', i), '_onedata_STBFH_MNT.mat']);
    new_1=[];
    new_2=[];
    for j=1:80
       if(allgoodevents(1,j)==1)
           if(allevents(1,j)==13)
               new_1=[new_1,j];
           end
           else
               new_2=[new_2,j];
       end
    end
    new_psd_1=zeros(501,24,128,size(new_1,2));
    new_psd_2=zeros(501,24,128,size(new_2,2));
    for j=1:size(new_1,2)
        new_psd_1(:,:,:,j)=psd(:,:,:,new_1(j));
    end
    for j=1:size(new_2,2)
        new_psd_2(:,:,:,j)=psd(:,:,:,new_2(j));
    end
    mean_psd_1=mean(new_psd_1,4);   %Face
    mean_psd_2=mean(new_psd_2,4);   %House
        for k=1:24
            psdInTheta = reshape(mean_psd_1(5:8,k,:),[4,128]);
            [eachMax,eachCh] = max(psdInTheta,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            theta(1,k,1,i) = bandMax;
            theta(2,k,1,i) = maxCh;
            psdInAlpha = reshape(mean_psd_1(9:13,k,:),[5,128]);
            [eachMax,eachCh] = max(psdInAlpha,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            alpha(1,k,1,i) = bandMax;
            alpha(2,k,1,i) = maxCh;
            psdInLowbeata = reshape(mean_psd_1(15:21,k,:),[7,128]);
            [eachMax,eachCh] = max(psdInLowbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowbeta(1,k,1,i) = bandMax;
            lowbeta(2,k,1,i) = maxCh;
            psdInHighbeata = reshape(mean_psd_1(21:31,k,:),[11,128]);
            [eachMax,eachCh] = max(psdInHighbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highbeta(1,k,1,i) = bandMax;
            highbeta(2,k,1,i) = maxCh;
            psdInLowgamma = reshape(mean_psd_1(31:61,k,:),[31,128]);
            [eachMax,eachCh] = max(psdInLowgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowgamma(1,k,1,i) = bandMax;
            lowgamma(2,k,1,i) = maxCh;
            psdInHighgamma = reshape(mean_psd_1(61:101,k,:),[41,128]);
            [eachMax,eachCh] = max(psdInHighgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highgamma(1,k,1,i) = bandMax;
            highgamma(2,k,1,i) = maxCh;
        end
        for k=1:24
            psdInTheta = reshape(mean_psd_2(5:8,k,:),[4,128]);
            [eachMax,eachCh] = max(psdInTheta,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            theta(1,k,2,i) = bandMax;
            theta(2,k,2,i) = maxCh;
            psdInAlpha = reshape(mean_psd_2(9:13,k,:),[5,128]);
            [eachMax,eachCh] = max(psdInAlpha,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            alpha(1,k,2,i) = bandMax;
            alpha(2,k,2,i) = maxCh;
            psdInLowbeata = reshape(mean_psd_2(15:21,k,:),[7,128]);
            [eachMax,eachCh] = max(psdInLowbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowbeta(1,k,2,i) = bandMax;
            lowbeta(2,k,2,i) = maxCh;
            psdInHighbeata = reshape(mean_psd_2(21:31,k,:),[11,128]);
            [eachMax,eachCh] = max(psdInHighbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highbeta(1,k,2,i) = bandMax;
            highbeta(2,k,2,i) = maxCh;
            psdInLowgamma = reshape(mean_psd_2(31:61,k,:),[31,128]);
            [eachMax,eachCh] = max(psdInLowgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowgamma(1,k,2,i) = bandMax;
            lowgamma(2,k,2,i) = maxCh;
            psdInHighgamma = reshape(mean_psd_2(61:101,k,:),[41,128]);
            [eachMax,eachCh] = max(psdInHighgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highgamma(1,k,2,i) = bandMax;
            highgamma(2,k,2,i) = maxCh;
        end
end
for i=1:20
    load(['stft_VP', sprintf('%02d', i), '_psd_con2.mat']);
    load(['CLAS_VP', sprintf('%02d', i), '_onedata_STBFH_MNT.mat']);
    new_1=[];
    new_2=[];
    for j=1:80
       if(allgoodevents(2,j)==1)
           if(allevents(2,j)==13)
               new_1=[new_1,j];
           end
           else
               new_2=[new_2,j];
       end
    end
    new_psd_1=zeros(501,24,128,size(new_1,2));
    new_psd_2=zeros(501,24,128,size(new_2,2));
    for j=1:size(new_1,2)
        new_psd_1(:,:,:,j)=psd(:,:,:,new_1(j));
    end
    for j=1:size(new_2,2)
        new_psd_2(:,:,:,j)=psd(:,:,:,new_2(j));
    end
    mean_psd_1=mean(new_psd_1,4);   %digit
    mean_psd_2=mean(new_psd_2,4);   %letter
        for k=1:24
            psdInTheta = reshape(mean_psd_1(5:8,k,:),[4,128]);
            [eachMax,eachCh] = max(psdInTheta,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            theta(1,k,3,i) = bandMax;
            theta(2,k,3,i) = maxCh;
            psdInAlpha = reshape(mean_psd_1(9:13,k,:),[5,128]);
            [eachMax,eachCh] = max(psdInAlpha,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            alpha(1,k,3,i) = bandMax;
            alpha(2,k,3,i) = maxCh;
            psdInLowbeata = reshape(mean_psd_1(15:21,k,:),[7,128]);
            [eachMax,eachCh] = max(psdInLowbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowbeta(1,k,3,i) = bandMax;
            lowbeta(2,k,3,i) = maxCh;
            psdInHighbeata = reshape(mean_psd_1(21:31,k,:),[11,128]);
            [eachMax,eachCh] = max(psdInHighbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highbeta(1,k,3,i) = bandMax;
            highbeta(2,k,3,i) = maxCh;
            psdInLowgamma = reshape(mean_psd_1(31:61,k,:),[31,128]);
            [eachMax,eachCh] = max(psdInLowgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowgamma(1,k,3,i) = bandMax;
            lowgamma(2,k,3,i) = maxCh;
            psdInHighgamma = reshape(mean_psd_1(61:101,k,:),[41,128]);
            [eachMax,eachCh] = max(psdInHighgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highgamma(1,k,3,i) = bandMax;
            highgamma(2,k,3,i) = maxCh;
        end
        for k=1:24
            psdInTheta = reshape(mean_psd_2(5:8,k,:),[4,128]);
            [eachMax,eachCh] = max(psdInTheta,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            theta(1,k,4,i) = bandMax;
            theta(2,k,4,i) = maxCh;
            psdInAlpha = reshape(mean_psd_2(9:13,k,:),[5,128]);
            [eachMax,eachCh] = max(psdInAlpha,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            alpha(1,k,4,i) = bandMax;
            alpha(2,k,4,i) = maxCh;
            psdInLowbeata = reshape(mean_psd_2(15:21,k,:),[7,128]);
            [eachMax,eachCh] = max(psdInLowbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowbeta(1,k,4,i) = bandMax;
            lowbeta(2,k,4,i) = maxCh;
            psdInHighbeata = reshape(mean_psd_2(21:31,k,:),[11,128]);
            [eachMax,eachCh] = max(psdInHighbeata,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highbeta(1,k,4,i) = bandMax;
            highbeta(2,k,4,i) = maxCh;
            psdInLowgamma = reshape(mean_psd_2(31:61,k,:),[31,128]);
            [eachMax,eachCh] = max(psdInLowgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            lowgamma(1,k,4,i) = bandMax;
            lowgamma(2,k,4,i) = maxCh;
            psdInHighgamma = reshape(mean_psd_2(61:101,k,:),[41,128]);
            [eachMax,eachCh] = max(psdInHighgamma,[],2);
            [bandMax,index] = max(eachMax);
            maxCh = eachCh(index);
            highgamma(1,k,4,i) = bandMax;
            highgamma(2,k,4,i) = maxCh;
        end
end
save('Channels_mean_max_psd.mat','theta','alpha','lowbeta','highbeta','lowgamma','highgamma')